tools/spy-detector/src/main.rs

```diff
@@ -0,0 +1,262 @@
+struct Args {
+    /// set the threshold for spy detection (default value is 5)
+    #[arg(short, long, default_value = "5")]
+    threshold: u32,
+
+    /// The log level the tool should run with. Valid log levels
+    /// are "trace", "debug", "info", "warn", "error". See https://docs.rs/log/latest/log/enum.Level.html
+    #[arg(short, long, default_value_t = log::Level::Debug)]
+    log_level: log::Level,
+}
+
+#[derive(Debug, Default)]
+struct PeerStats {
+    inv_tx_received: AtomicU32,            // TX(INV) received by the peer
+    inv_tx_sent: AtomicU32,                // TX(INV) sent by the peer
+    inv_wtx_sent: AtomicU32,               // WTX(INV) received by the peer
+    inv_wtx_received: AtomicU32,           // WTX(INV) received by the peer
+    inv_witnesstx_received: AtomicU32,     // WitnessTX(INV) received by the peer
+    inv_witnesstx_sent: AtomicU32,         // WitnessTX(INV) sent by the peer
+    getdata_witnesstx_sent: AtomicU32,     // WitnessTX(GETDATA) sent by the peer
+    getdata_witnesstx_received: AtomicU32, // WitnessTX(GETDATA) received by the peer
+    tx_sent: AtomicU32,                    // TX sent by the peer
+    tx_received: AtomicU32,                // TX received by the peer
+    last_activity: Option<Instant>,        // Last activity time of the peer
+}
+
+type PeerMap = Arc<DashMap<String, PeerStats>>;
+
+fn main() {
+    let args = Args::parse();
+
+    //to-do: use threshold at appropriate location
+    let threshold = &args.threshold;
+
+    simple_logger::init_with_level(args.log_level).unwrap();
+
+    log::info!(target: LOG_TARGET, "Starting spy-detector...",);
+
+    log::info!(target: LOG_TARGET, "Starting spy-detector...",);
+
+    let sub = Socket::new(Protocol::Sub0).unwrap();
+    sub.dial(ADDRESS).unwrap();
+
+    let all_topics = vec![];
+    sub.set_opt::<Subscribe>(all_topics).unwrap();
+
+    let peer_map: PeerMap = Arc::new(DashMap::new());
+
+    // Spawn a thread for periodic stats display - 2 mins
+    let display_peer_map = peer_map.clone();
+    std::thread::spawn(move || {
+        let mut last_display = Instant::now();
+        loop {
+            let now = Instant::now();
+            if now.duration_since(last_display) >= STATS_INTERVAL {
+                display_all_stats(&display_peer_map);
+                last_display = now;
+            }
+            std::thread::sleep(Duration::from_secs(1));
+        }
+    });
+
+    log::info!(target: LOG_TARGET, "Spy-detector started",);
+
+    loop {
+        let msg = sub.recv().unwrap();
+        let message = event_msg::EventMsg::decode(msg.as_slice()).unwrap().event;
+
+        if let Some(event) = message {
+            match event {
+                Event::Msg(msg) => {
+                    let msg_type: u32;
+
+                    if msg.meta.inbound {
+                        msg_type = 0;
+                    } else {
+                        msg_type = 1;
+                    };
```

Comment:
Why do you re encode the `msg.meta.inbound` boolean as an `u32`? You only ever check if it's ` == 0`. This should stay a boolean!